File: gaussian_basis/gaussian_basis/molecular_geometry.py
from typing import Dict, List, Union
import numpy as np


def copy_position_list(positions: List[np.ndarray]) -> List[np.ndarray]:
    return [r.copy() for r in positions]


def copy_config(config: Dict[str, List[np.ndarray]]
                ) -> Dict[str, List[np.ndarray]]:
    return {k: copy_position_list(config[k]) for k in config}


def rotate(r, angle, rotation_axis, offset):
    def mult(a, b):
        return np.array([a[0]*b[0] - a[1]*b[1] - a[2]*b[2] - a[3]*b[3],
                         a[0]*b[1] + a[1]*b[0] + a[2]*b[3] - a[3]*b[2],
                         a[0]*b[2] + a[2]*b[0] + a[3]*b[1] - a[1]*b[3],
                         a[0]*b[3] + a[3]*b[0] + a[1]*b[2] - a[2]*b[1]])
    c, s = np.cos(angle/2.0), np.sin(-angle/2.0)
    norm = np.linalg.norm(rotation_axis)
    rot = np.array([c,
                    s*rotation_axis[0]/norm, 
                    s*rotation_axis[1]/norm, 
                    s*rotation_axis[2]/norm])
    rot_inv = np.array([rot[0], -rot[1], -rot[2], -rot[3]])
    q0 = np.array([1.0, 
                   r[0] - offset[0], r[1] - offset[1], r[2] - offset[2]])
    qf = mult(rot_inv, mult(q0, rot))
    return np.array([qf[1] + offset[0],
                     qf[2] + offset[1],
                     qf[3] + offset[2]])


def cross(a, b):
    # print(a.shape, b.shape)
    return np.array([a[1]*b[2] - a[2]*b[1],
                     -a[0]*b[2] + a[2]*b[0],
                     a[0]*b[1] - a[1]*b[0]])


class MolecularGeometry:

    config: Dict[str, List[np.ndarray]]

    def __init__(self, **kw):
        """MolecularGeometry constructor.

        If an argument is supplied, it must be the key-value pair

        config={<element>: <positions>, ...},

        where <element> is the atomic number of an atom and
        <positions> is a list of Numpy arrays of length 3
        where each numpy array gives the position for where
        this atom can be found.

        As an example,
        
        config={'H': [np.array([1.0, 0.0, 0.0]), np.array([0.0, 1.0, 0.0])],
                'O': [np.array([0.0, 0.0, 0.0])]}

        places hydrogen atoms at (1, 0, 0) and (0, 1, 0) and an oxygen
        atom at the origin.
        """
        self.config = {}
        if 'config' in kw:
            self.config = kw['config']
    
    def add_atom(self, atomic_symbol: 'str', position: np.ndarray):
        if atomic_symbol not in self.config:
            self.config[atomic_symbol] = []
        self.config[atomic_symbol].append(position)

    def __getitem__(self, atomic_symbol: str) -> np.ndarray:
        return np.array(self.config[atomic_symbol])

    def __add__(
            self,
            other: Union[np.ndarray, 'MolecularGeometry', int]
            ) -> 'MolecularGeometry':
        if isinstance(other, np.ndarray):
            config = copy_config(self.config)
            # print(config)
            for k in config:
                # print(config[k])
                for i in range(len(config[k])):
                    config[k][i] += other
            return MolecularGeometry(config=config)
        elif type(self) == type(other):
            config = copy_config(self.config)
            other_config = copy_config(other.config)
            for k in other_config:
                if k not in config:
                    config[k] = []
                config[k].extend(other_config[k])
            return MolecularGeometry(config=config)
        MolecularGeometry(config=copy_config(self.config))

    def __radd__(self, other: Union[np.ndarray, 'MolecularGeometry', int]
                 ) -> 'MolecularGeometry':
        return self.__add__(other)

    def __sub__(self, other: np.ndarray) -> 'MolecularGeometry':
        return self.__add__(-other)

    def __mul__(self, other: Union[float, np.ndarray]
                ) -> 'MolecularGeometry':
        config = copy_config(self.config)
        for k in config:
            for i in range(config[k]):
                config['k'][i] *= other
        return MolecularGeometry(config=config)
    
    def rotate(self, angle: float, rotation_axis: np.ndarray,
               offset: np.ndarray = np.zeros([3])
               ):
        for k in self.config:
            for i, r in enumerate(self.config[k]):
                self.config[k][i] = rotate(r, angle, rotation_axis, offset)

    def get_nuclear_configuration(self) -> List[List[Union[np.ndarray, int]]]:
        """ Get the configuration of the nuclear charges.

        It returns a list of the form

        [<position>: <charge>, ...],

        where <position> is a length 3 numpy array that indicates
        the position of the charge and <charge> is its charge strength.

        """
        nuc_config = []
        for k in self.config:
            if k == 'H':
                charge = 1
            elif k == 'C':
                charge = 6
            elif k == 'N':
                charge = 7
            elif k == 'O':
                charge = 8
            elif k == 'F':
                charge = 9
            else:
                raise NotImplementedError
            for i in range(len(self.config[k])):
                nuc_config.append([self.config[k][i].copy(), charge])
        return nuc_config


def make_oh(oh_length: float,
            axis=np.array([0.0, 0.0, 1.0])) -> 'MolecularGeometry':
    a = axis/np.linalg.norm(axis)
    geom = MolecularGeometry(config={
        'O': [np.array([0.0, 0.0, 0.0])],
        'H': [oh_length*a]
    })
    return geom


def make_co(co_length: float,
            axis=np.array([0.0, 0.0, 1.0])) -> 'MolecularGeometry':
    a = axis/np.linalg.norm(axis)
    geom = MolecularGeometry(config={
        'C': [np.array([0.0, 0.0, 0.0])],
        'O': [co_length*a]
    })
    return geom


def make_ch2(ch1: float, ch2: float,
             hch_angle: float,
             axis1: np.ndarray, axis2: np.ndarray
             ) -> 'MolecularGeometry':
    a1 = axis1/np.linalg.norm(axis1)
    a2 = axis2/np.linalg.norm(axis2)
    return MolecularGeometry(config={
        'C': [np.array([0.0, 0.0, 0.0])],
        'H': [ch1*(a1*np.cos(hch_angle/2.0) + a2*np.sin(hch_angle/2.0)), 
              ch2*(a1*np.cos(hch_angle/2.0) - a2*np.sin(hch_angle/2.0))],
    })


def make_ch3(ch1: float, ch2: float, ch3: float,
             ach_angle: float,
             axis1: np.ndarray, axis2: np.ndarray
             ) -> 'MolecularGeometry':
    a = axis1/np.linalg.norm(axis1)
    x = axis2/np.linalg.norm(axis2)
    y = cross(a, x)
    p1 = x
    phi = 2.0*np.pi/3.0
    p2 = x*np.cos(phi) + y*np.sin(phi)
    p3 = x*np.cos(2.0*phi) + y*np.sin(2.0*phi)
    return MolecularGeometry(config={
        'C': [np.array([0.0, 0.0, 0.0])],
        'H': [ch1*(a*np.cos(ach_angle) + p1*np.sin(ach_angle)),
              ch2*(a*np.cos(ach_angle) + p2*np.sin(ach_angle)),
              ch3*(a*np.cos(ach_angle) + p3*np.sin(ach_angle))],
    })


if __name__ == '__main__':

    import matplotlib.pyplot as plt

    # config = {'H': [np.array([1.0, 0.0, 0.0]),
    #                 np.array([-1.0, 0.0, 0.0])],
    #           'C': [np.array([0.0, 0.0, 0.0])]}
    # config2 = copy_config(config)
    # config2['H'].append(np.array([0.0, 1.0, 0.0]))
    # config2['C'][0][2] = -0.4
    # geom = MolecularGeometry(config=config2)
    # print(config2)
    # geom.rotate(np.pi/4.0, np.array([0.0, 0.0, 1.0]))
    # print(geom.config)

    geom1 = make_ch2(1.5, 1.5, 3.0 * np.pi / 2.0,
                     np.array([1.0, 0.0, 0.0]),
                     np.array([0.0, 0.0, 1.0]))
    plt.scatter(geom1['C'].T[0], geom1['C'].T[2])
    plt.scatter(geom1['H'].T[0], geom1['H'].T[2])
    plt.show()
    plt.close()

    geom2 = make_ch3(2.0, 1.5, 1.5, np.pi/4.0,
                    np.array([1.0, 0.0, 0.0]),
                    np.array([0.0, 0.0, 1.0]))

    fig = plt.figure()
    ax = fig.add_subplot(projection='3d')
    ax.scatter(geom2['H'].T[0],
               geom2['H'].T[1],
               geom2['H'].T[2])
    ax.scatter(geom2['C'].T[0],
               geom2['C'].T[1],
               geom2['C'].T[2])
    ax.set_xlabel('x')
    ax.set_ylabel('y')
    ax.set_zlabel('z')
    plt.show()
    plt.close()

File: gaussian_basis/gaussian_basis/src2/molecular_geometry.cpp
#include "molecular_geometry.hpp"
#include <cmath>

using namespace molecular_geometry;


static spatial::Vector rotate(
    spatial::Vector r, double angle, 
    spatial::Vector axis, spatial::Vector offset) {
    #define MUL(a, b)\
        {a[0]*b[0] - a[1]*b[1] - a[2]*b[2] - a[3]*b[3],\
        a[0]*b[1] + a[1]*b[0] + a[2]*b[3] - a[3]*b[2],\
        a[0]*b[2] + a[2]*b[0] + a[3]*b[1] - a[1]*b[3],\
        a[0]*b[3] + a[3]*b[0] + a[1]*b[2] - a[2]*b[1]} 
    double c = cos(angle/2.0), s = sin(-angle/2.0);
    double norm = sqrt(
        axis.x*axis.x + axis.y*axis.y + axis.z*axis.z);
    double rot[4] = {c, s*axis.x/norm, s*axis.y/norm, s*axis.z/norm};
    double rot_inv[4] = {rot[0], -rot[1], -rot[2], -rot[3]};
    double q0[4] = {1.0, 
        r[0] - offset[0], r[1] - offset[1], r[2] - offset[2]};
    double q0_rot[4] = MUL(q0, rot);
    double qf[4] = MUL(rot_inv, q0_rot);
    #undef MUL
    return spatial::Vector{
        .t=0.0,
        .x=qf[1] + offset[0],
        .y=qf[2] + offset[1],
        .z=qf[3] + offset[2]};
}


void MolecularGeometry::create_empty_m_positions() {
    m_positions = {{}};
    // Only goes up to Calcium, for now.
    for (int _ = 1; _ <= 20; _++)
        m_positions.push_back({});
}

MolecularGeometry::MolecularGeometry() {
    this->create_empty_m_positions();
}

MolecularGeometry::MolecularGeometry(
    AtomicSymbol atomic_symbol,
    const spatial::Vector &position
) {
    this->create_empty_m_positions();
    m_positions[int(atomic_symbol)].push_back(position);
}

MolecularGeometry::MolecularGeometry(
    AtomicSymbol atomic_symbol, 
    const std::vector<spatial::Vector> &positions
) {
    this->create_empty_m_positions();
    for (auto &position: positions)
        m_positions[int(atomic_symbol)].push_back(position);
}

MolecularGeometry::MolecularGeometry(
    std::vector<AtomicSymbol> atomic_symbols, 
    const std::vector<std::vector<spatial::Vector>> &positions_list) {
    this->create_empty_m_positions();
    int i = 0;
    for (AtomicSymbol &atomic_symbol: atomic_symbols) {
        for (auto &position: positions_list[i])
            m_positions[int(atomic_symbol)].push_back(position);
        i++;
    }
}

MolecularGeometry
::MolecularGeometry(const std::map<AtomicSymbol, 
    std::vector<spatial::Vector>> &atom_positions) {
    for (auto &e: atom_positions) {
        AtomicSymbol atomic_symbol = e.first;
        std::vector<spatial::Vector> positions = e.second;
        for (auto &position: positions)
            m_positions[int(atomic_symbol)].push_back(position);
    }
}

void MolecularGeometry
::add_atom(AtomicSymbol atomic_symbol, const spatial::Vector &position) {
    m_positions[int(atomic_symbol)].push_back(position);
}

std::vector<spatial::Vector> MolecularGeometry
::operator[](AtomicSymbol atomic_symbol) const {
    return m_positions[int(atomic_symbol)];
}

MolecularGeometry MolecularGeometry
::operator+(const spatial::Vector &translate_position) const {
    MolecularGeometry res{};
    for (int n = 1; n < m_positions.size(); n++) {
        for (auto &position: m_positions[n])
            res.m_positions[n].push_back(
                position + translate_position);
    }
    return res;
}

MolecularGeometry MolecularGeometry
::operator-(const spatial::Vector &translate_position) const {
    MolecularGeometry res{};
    for (int n = 1; n < m_positions.size(); n++) {
        for (auto &position: m_positions[n])
            res.m_positions[n].push_back(
                position - translate_position);
    }
    return res;
}

MolecularGeometry MolecularGeometry
::operator+(const MolecularGeometry &geom) const {
    MolecularGeometry res{};
    for (int n = 1; n < m_positions.size(); n++) {
        for (auto &position: m_positions[n])
            res.m_positions[n].push_back(position);
        for (auto &position: geom.m_positions[n])
            res.m_positions[n].push_back(position);
    }
    return res;
}

MolecularGeometry MolecularGeometry
::rotate(
    double angle, 
    const spatial::Vector &rotation_axis, 
    const spatial::Vector &offset) const {
    MolecularGeometry res{};
    for (int n = 1; n < m_positions.size(); n++) {
        for (auto &position: m_positions[n])
            res.m_positions[n].push_back(
                ::rotate(position, angle, rotation_axis, offset)
            );
    }
    return res;
}

NuclearChargesArray MolecularGeometry
::get_nuclear_configuration() const {
    NuclearChargesArray config({});
    for (int atomic_number = 1; 
        atomic_number < m_positions.size(); atomic_number++) {
        for (auto &position: m_positions[atomic_number])
            config.push_back({
                .position=position,
                .strength=atomic_number
            });
    }
    return config;
}


molecular_geometry::MolecularGeometry make_oh(
    double oh_length, const spatial::Vector &axis
) {
    return MolecularGeometry({
        {AtomicSymbol::H, {oh_length*axis}},
        {AtomicSymbol::O, {{.t=0.0, .x=0.0, .y=0.0, .z=0.0}}},
    });
}

molecular_geometry::MolecularGeometry
    make_co(double co_length, const spatial::Vector &axis=spatial::Vector{.x=0.0, .y=0.0, .z=1.0}) {
    return MolecularGeometry ({
        {AtomicSymbol::C, {{.t=0.0, .x=0.0, .y=0.0, .z=0.0}}},
        {AtomicSymbol::O, {co_length*axis}},
    });
}

molecular_geometry::MolecularGeometry
make_ch2(double ch1, double ch2, double hch_angle,
    const spatial::Vector &axis1, const spatial::Vector &axis2) {
    spatial::Vector a1{axis1}, a2{axis2};
    return MolecularGeometry(
        {
            {AtomicSymbol::C, {{.t=0.0, .x=0.0, .y=0.0, .z=0.0}}},
            {AtomicSymbol::H, 
                {
                    ch1*(a1*cos(hch_angle/2.0) + a2*sin(hch_angle/2.0)),
                    ch2*(a1*cos(hch_angle/2.0) - a2*sin(hch_angle/2.0))
                }
            }
        }
    );
}

molecular_geometry::MolecularGeometry
make_ch3(double ch1, double ch2, double ch3, double ach_angle,
    const spatial::Vector &axis1, const spatial::Vector &axis2) {
    spatial::Vector a{axis1}, x{axis2};
    spatial::Vector y = cross(a, x);
    spatial::Vector p1 = x;
    double phi = 2.0*(3.141592653589793)/3.0;
    spatial::Vector p2 = x*cos(phi) + y*sin(phi);
    spatial::Vector p3 = x*cos(2.0*phi) + y*sin(2.0*phi);
    return MolecularGeometry(
        {
            {AtomicSymbol::C, {{.t=0.0, .x=0.0, .y=0.0, .z=0.0}}},
            {AtomicSymbol::H, {
                {ch1*(a*cos(ach_angle) + p1*sin(ach_angle))},
                {ch2*(a*cos(ach_angle) + p2*sin(ach_angle))},
                {ch3*(a*cos(ach_angle) + p3*sin(ach_angle))}
            }}
        }
    );
}
